### DISCUSSION TITLE
Small favor if possible

### DISCUSSION THREAD

**Discussion by ErikssonPer** [ai-republic/bms-to-inverter#45] [Q&A]
Hi,
Greetings from Sweden.
Found this page recently and really liked what I read here, it seems just like what I need for my small project at home.

I have a Growatt SPH10000TL3 BH-UP coupled to a self built 33kWh high voltage battery with it's own BMS.
Basically 100 lifepo4 cells connected in series with a bms.
The SPH is in "lead-acid" battery mode which works ok but I'd really like to have the BMS talking to it to make the system work better.
I also have an ARK battery but that's not used now since it's too small for my needs but I decided to keep it since the resale value isn't great and it might come in handy some day.

When I found this project I went ahead and reconnected the ARK battery and tapped into the can between it and the inverter with an rpi4 with 2ch canhat, dumped some candata from it which worked fine, It matches what I've read here for growatt sph systems.

My next step was to attempt to dump candata from the other BMS too see if the data needed for bmstoinverter is present but that seem to be to advanced for me, I'm a builder guy, not a programmer guy.......so here I need some help to get me started.
It's a cheap chinese bms and the documentation isn't the best and the support isn't much help either.
What I got is a pdf documenting their CAN protocol and they told me I need to send a handshake message first and then the bms will start sending data back but I don't understand how that message should look.
I was hoping I could just send it with  cansend and then get data back I could play with.

I've attached the file if anyone would like to take a look.
It's the last message on the last page that's the handshake.

/Per

**Comment by ErikssonPer** [comment]
[TopBMS CAN protocol.pdf](https://github.com/user-attachments/files/16475442/TopBMS.CAN.protocol.pdf)

**Comment by ai-republic** [comment]
Hmm, not so clear in the specification about the 'CAN Receive' part ;)
Have you tried:
----- id ---- -L- --flags-- -C--------- data --------- (L=length, C=command)
4F 00 00 00 08 00 00 00 10 00 00 00 00 00 00 02

**Reply by ErikssonPer** [reply]
Success! 
I'm recieving can data from it now.
Attached a txtfile with a sample.
Most values have to be divided by 10 or 100 to get the actual value.

Would it be possible to use the data from this bms and run it through your program to the sph inverter?
[Candump from topbms.txt](https://github.com/user-attachments/files/16483102/Candump.from.topbms.txt)

**Comment by ai-republic** [comment]
Alright, cool! 
So it was the second command (0x1C) in the specification to wake up the BMS or did you have to send both (command 0x10 and another with 0x1C)? Or did the BMS reply to command 0x10 with the reply 0x1C?

Yes, the specification on the 'CAN send' is quite clear with the 0xF5 start flag and their command flags and interpretation. So implementing that binding will be no problem.

**Comment by ErikssonPer** [comment]
Only needed 0x1C
I did a google translate on the chinese version of the document and there it said that 0x10 was for turning on/off bluetooth communication.

I'm on vacation now but will take some more dumps while charging, discharging and disconnected to see if it matches what the document says when I'm back home.

**Comment by ai-republic** [comment]
Ok, let me know if you need any help 😄

**Comment by ai-republic** [comment]
Guess this is answered successfully :)

**Comment by ErikssonPer** [comment]
Yes, I did do some checking of the candata and had most of it figured out but some other things with it irritaded me so I decided to go with an EMUSbms.com instead, that has way better documenttion of it's can protcol and also out of the box have the ability to emulate several other protocols which should make life easier hopefully.
I'll get back when it's installed and running.
